### PR TITLE
Resource table

### PR DESCRIPTION
**Pull request by rasbt**
I made a resource table that I think will be helpful for people who have questions about the resource requirements. Also, it could be a useful reference for spot checking new select PRs to see whether they improve/decrease performance and memory requirements.

One question though is regarding the runtime of the multi-GPU cases. If I run `max_iters = 1000` on 1 GPU, it will do 1000 iterations. If I run the same code on 2 GPUs, is it actually iterating through 2000 examples (even though it prints 1000 iterations)? 

EDIT: I want to add 8-GPU tests some time too but that's currently not possible due to the 4 GPUs currently being occupied for other things)

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

**Comment by rasbt** [comment]
Added some additional info about the hardware and relevant libraries @carmocca

File: tutorials/finetune_lora.md [hunk 1]
```diff
@@ -57,7 +58,7 @@ and optionally with double-quantization:
 python finetune/lora.py --quantize "bnb.nf4-dq"
 ```
 
-The table below lists a comparison with different settings on a StableLM 3B model finetuned with LoRA on Alpaca for 5,000 iterations using a microbatch size of 4:
```
Comment (carmocca) [inline tutorials/finetune_lora.md:60]: Was 5k a mistake? I would expect that changing the # of iters requires updating the training time and loss in the tables

File: tutorials/resource-tables.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+|       |               |              |                 |                      |          |             |                    |                              |
+| 13 B  | Llama 2       | None         | 1               | 6,553,600            | 2 x A100 | OOM         | -                  | -                            |
+| 13 B  | Llama 2       | bnb.nf4      | 1               | 6,553,600            | 2 x A100 | N/A         | -                  | -                            |
+| 13 B  | Llama 2       | bnb.nf4-dq   | 1               | 6,553,600            | 2 x A100 | N/A         | -                  | -                            |
+|       |               |              |                 |                      |          |             |                    |                              |
+| 13 B  | Llama 2       | None         | 1               | 6,553,600            | 4 x A100 | 35.57 GB    | 10.25 min          | 512.5 min                    |
+| 40 B  | Falcon        | None         | 1               | 12,042,240           | 4 x A100 | OOM         | -                  | -                            |
+
```
Comment (carmocca) [inline tutorials/resource-tables.md:100]: I think you forgot adding nbsp here

File: tutorials/resource-tables.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+|-------|---------------|--------------|----------|-------------------------------------------|-----------|
+| 1.3 B | phi-1.5       | None         | 1 x A100 | 2.86 GB                                   | 42.56     |
+| 1.3 B | phi-1.5       | bnb.nf4      | 1 x A100 | 1.39 GB                                   | 22.89     |
+| 1.3 B | phi-1.5       | bnb.nf4-dq   | 1 x A100 | 1.33 GB                                   | 22.75     |
+| 1.3 B | phi-1.5       | gptq.int4    | 1 x A100 | 1.16 GB                                   | 6.51      |
+|       |               |              |          |                                           |           |
+| 3 B   | StabeLM Alpha | None         | 1 x A100 | 7.30 GB                                   | 49.01     |
+| 3 B   | StabeLM Alpha | bnb.nf4      | 1 x A100 | 3.20 GB                                   | 29.04     |
+| 3 B   | StabeLM Alpha | bnb.nf4-dq   | 1 x A100 | 3.04 GB                                   | 27.15     |
+| 3 B   | StabeLM Alpha | gptq.int4    | 1 x A100 | 2.43 GB                                   | 5.9       |
```
Comment (carmocca) [inline tutorials/resource-tables.md:113]: Typo: StableLM

File: tutorials/resource-tables.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+|       |               |              |          |                                           |           |
+| 13 B  | Llama 2       | None         | 1 x A100 | 26.21 GB                                  | 24.82     |
+| 13 B  | Llama 2       | bnb.nf4      | 1 x A100 | 8.32 GB                                   | 16.73     |
+| 13 B  | Llama 2       | bnb.nf4-dq   | 1 x A100 | 7.72 GB                                   | 14.43     |
+| 13 B  | Llama 2       | gptq.int4    | 1 x A100 | 7.14 GB                                   | 4.17      |
+|       |               |              |          |                                           |           |
+| 34 B  | CodeLlama     | None         | 1 x A100 | OOM                                       | -         |
+| 34 B  | CodeLlama     | bnb.nf4      | 1 x A100 | 20.52 GB                                  | 14.32     |
+| 34 B  | CodeLlama     | bnb.nf4-dq   | 1 x A100 | 18.95 GB                                  | 12.37     |
+| 34 B  | CodeLlama     | gptq.int4    | 1 x A100 | OOM (quantize script)                     | -         |
```
Comment (carmocca) [inline tutorials/resource-tables.md:128]: What do you mean with "Quantize script" here?

File: tutorials/resource-tables.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+|       |               |              |          |                                           |           |
+| 40 B  | Falcon        | None         | 1 x A100 | OOM                                       | -         |
+| 40 B  | Falcon        | bnb.nf4      | 1 x A100 | 26.55 GB                                  | 13.25     |
+| 40 B  | Falcon        | bnb.nf4-dq   | 1 x A100 | 24.63 GB                                  | 11.64     |
+| 40 B  | Falcon        | gptq.int4    | 1 x A100 | OOM (quantize script)                     | -         |
+|       |               |              |          |                                           |           |
+| 70 B  | Llama 2       | None         | 1 x A100 | OOM                                       | -         |
+| 70 B  | Llama 2       | bnb.nf4      | 1 x A100 | CUDA error: CUBLAS_STATUS_NOT_INITIALIZED | -         |
```
Comment (carmocca) [inline tutorials/resource-tables.md:136]: Interesting. Can you paste the stacktrace in a comment here for future reference?

File: tutorials/resource-tables.md (hunk 4 above)
Comment (rasbt) [reply tutorials/resource-tables.md:128]: For GPTQ.int4, it's a 2-step procedure where you have to quantize the model before you run it. So here, this means the script that is quantizing the model, not using the model for inference.

File: tutorials/resource-tables.md (hunk 3 above)
Comment (rasbt) [reply tutorials/resource-tables.md:113]: good catch!

**Comment by rasbt** [comment]
> Interesting. Can you paste the stacktrace in a comment here for future reference?

@carmocca It's a weird one, I haven't seen it with other models and can reproduce:

```
../aten/src/ATen/native/cuda/Indexing.cu:1292: indexSelectLargeIndex: block: [623,0,0], thread: [28,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1292: indexSelectLargeIndex: block: [623,0,0], thread: [29,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1292: indexSelectLargeIndex: block: [623,0,0], thread: [30,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
...
...
...
../aten/src/ATen/native/cuda/Indexing.cu:1292: indexSelectLargeIndex: block: [666,0,0], thread: [125,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1292: indexSelectLargeIndex: block: [666,0,0], thread: [126,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1292: indexSelectLargeIndex: block: [666,0,0], thread: [127,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
Traceback (most recent call last):
  File "/data/sebastian/resource-table/lit-gpt/finetune/lora.py", line 335, in <module>
    CLI(setup)
  File "/home/sebastian/.local/lib/python3.10/site-packages/jsonargparse/_cli.py", line 85, in CLI
    return _run_component(component, cfg_init)
  File "/home/sebastian/.local/lib/python3.10/site-packages/jsonargparse/_cli.py", line 147, in _run_component
    return component(**cfg)
  File "/data/sebastian/resource-table/lit-gpt/finetune/lora.py", line 96, in setup
    fabric.launch(main, data_dir, checkpoint_dir, out_dir, quantize)
  File "/home/sebastian/.local/lib/python3.10/site-packages/lightning/fabric/fabric.py", line 834, in launch
    return self._wrap_and_launch(function, self, *args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/lightning/fabric/fabric.py", line 920, in _wrap_and_launch
    return to_run(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/lightning/fabric/fabric.py", line 925, in _wrap_with_setup
    return to_run(*args, **kwargs)
  File "/data/sebastian/resource-table/lit-gpt/finetune/lora.py", line 153, in main
    train(fabric, model, optimizer, scheduler, train_data, val_data, checkpoint_dir, out_dir, speed_monitor)
  File "/data/sebastian/resource-table/lit-gpt/finetune/lora.py", line 182, in train
    validate(fabric, model, val_data, tokenizer)  # sanity check
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/sebastian/resource-table/lit-gpt/finetune/lora.py", line 264, in validate
    logits = model(input_ids)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/lightning/fabric/wrappers.py", line 121, in forward
    output = self._forward_module(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/sebastian/resource-table/lit-gpt/lit_gpt/lora.py", line 498, in forward
    x = block(x, cos, sin, mask, input_pos)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/sebastian/resource-table/lit-gpt/lit_gpt/model.py", line 158, in forward
    h = self.attn(n_1, cos, sin, mask, input_pos)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/sebastian/resource-table/lit-gpt/lit_gpt/model.py", line 196, in forward
    qkv = self.attn(x)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/sebastian/resource-table/lit-gpt/lit_gpt/lora.py", line 372, in forward
    pretrained = self.linear(x)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sebastian/.local/lib/python3.10/site-packages/bitsandbytes/nn/modules.py", line 248, in forward
    out = bnb.matmul_4bit(x, self.weight.t(), bias=bias, quant_state=self.weight.quant_state)
  File "/home/sebastian/.local/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py", line 579, in matmul_4bit
    return MatMul4Bit.apply(A, B, out, bias, quant_state)
  File "/home/sebastian/.local/lib/python3.10/site-packages/torch/autograd/function.py", line 539, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/home/sebastian/.local/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py", line 516, in forward
    output = torch.nn.functional.linear(A, F.dequantize_4bit(B, state).to(A.dtype).t(), bias)
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`
```

File: tutorials/resource-tables.md (hunk 5 above)
Comment (rasbt) [reply tutorials/resource-tables.md:138]: Added it below so it doesn't get lost here.

**Comment by carmocca** [comment]
Thanks. Since it's an error in bitsandbytes, I wouldn't worry too much about it.

File: tutorials/finetune_lora.md (hunk 1 above)
Comment (rasbt) [reply tutorials/finetune_lora.md:60]: This was still the old table, but I just updated it to numbers from the large (new) table.

**Comment by rasbt** [comment]
> Thanks. Since it's an error in bitsandbytes, I wouldn't worry too much about it.

Agreed. It was just odd that it occurred only with one of the models.

**Pull request merged**